proposals/0370-modifiers.rst

```diff
@@ -122,39 +126,136 @@ Proposed Change Specification
     have a known type if declared with a type signature. Alternatively, the
     modifier may have a top-level type signature.
 
-12. A modifier of type ``Multiplicity`` changes the multiplicity of the following arrow,
-    or following pattern-bound variable of a lambda,
-    or preceding record field.
-    Multiple modifiers of type ``Multiplicity`` on the same arrow are not allowed.
-    Any other use of a modifier is an error.
-
-13. ``-XLinearTypes`` implies ``-XModifiers``.
-
-14. Future modifiers will be put *before* the element they modify. Alternatively,
+12. Future modifiers will be put *before* the element they modify. Alternatively,
     a modifier can be put directly before a syntactic closer or separator, such
     as ``;`` or ``where`` or ``)``.
 
-15. Modifiers with an unknown meaning produce a warning, controlled by
-    ``-Wunknown-modifiers``. They are otherwise ignored. (However, in order to
-    know that a modifier is unknown, it still must be parsed, renamed, and type-checked.)
+13. Modifiers of unknown kind produce an error.
+
+14. Modifiers of known kind but with an unknown meaning produce a warning,
+    controlled by ``-Wunknown-modifiers``. They are otherwise ignored. (However,
+    in order to know that a modifier is unknown, it still must be parsed,
+    renamed, and type-checked.)
+
+LinearTypes
+-----------
+With ``-XLinearTypes``:
+
+* A modifier of type ``Multiplicity`` changes the multiplicity of the following
+  arrow, or following pattern-bound variable of a lambda, or following let or
+  where binding, or preceding record field. Multiple modifiers of type
+  ``Multiplicity`` on the same arrow are not allowed. Any other use of a
+  modifier still has no meaning.
+
+* The ``%1`` modifier is handled as a special case. It's renamed (and so
+  typechecked) the same as ``%One``, even if it appears somewhere that linear
+  modifiers aren't expected. If a user does want the modifier ``1 :: Nat``, they
+  can write it as ``%01``. (Requires ``-XDataKinds``.)
+
+* The linear arrow ``a ⊸ b`` has the same meaning as ```a %1 -> b``. Other
+  modifiers are accepted: ``a %Matchable ⊸ b`` has the same meaning as
+  ``a %Matchable %1 -> b``.
+
+With ``-XNoLinearTypes``, the ``%1`` modifier is not special. It refers to the
+type ``1 :: Nat`` and requires ``-XDataKinds``.
+
+``-XLinearTypes`` implies ``-XModifiers``. But the latter can be explicitly
```

Comment:
> Though I imagine others will disagree, I favor 3.

Indeed, I would oppose that. `Modifiers` would be a new extension, and we should mark it "Experimental", and we can't impose experimental extensions on users by default. If, on the other hand, we mark it "Stable", that would create obstacles to making changes to this new extension.

Personally I think there are plenty of use cases that modifiers could cover (e.g. replace various instance pragmas) that we should specify and implement before `-XModifiers` fossilizes and gets enabled by default.

Reply:
My thought is that we should feel free to add more functionality to `-XModifiers` even if it is enabled by default -- as long as all the changes are backward compatible. But actually we might decide we made a mistake somewhere and *remove* functionality from `-XModifiers`, which suggests it should remain off by default for now. So I think I favor (1), of my three options above: don't have `-XLinearTypes` and `-XModifiers` interact directly, but allow the `%` syntax with either `-XLinearTypes` *or* `-XModifiers` (though with only the latter, it would cause an error later).